(WIP) Add Protobuf schemas for existing PySyft messages and types

karlhigley opened the pull request:
A first stab at capturing the PySyft protocol in Protobuf, along the
lines of the ["Standardizing PySyft Protocol" RFC](https://github.com/OpenMined/rfcs/blob/master/20190822-standardizing-pysyft-protocol.md). The approach taken
here is aimed at transmitting Protobuf over Websockets, and doesn't
attempt to define gRPC services--see the [RFC PR discussion](https://github.com/OpenMined/rfcs/pull/6) for the
reasoning behind that. See also proposed changes 5-6 in the [serde
refactoring issue](https://github.com/OpenMined/PySyft/issues/2654).

This PR includes schemas for the existing messages in [message.py](https://github.com/OpenMined/PySyft/blob/master/syft/messaging/message.py) and
defines types where necessary to capture the content of those messages.
It includes a schema for Operation, which hopefully will be sufficiently
flexible to cover nearly all of the current uses of Operation in the PySyft
test suite. (I enumerated those uses by adding logging in Operation's
`__init__` and capturing the argument types. One known case that isn't
covered is nested tuples in the args, which I haven't sorted out yet.)

Schemas for Tensorflow types will be broadly similar and will make heavy
use of existing framework serialization (like the Torch types here do.) I
plan to add those in a future commit.

karlhigley committed:
(WIP) Add Protobuf schemas for existing PySyft messages and types

A first stab at capturing the PySyft protocol in Protobuf, along the
lines of the ["Standardizing PySyft Protocol" RFC](https://github.com/OpenMined/rfcs/blob/master/20190822-standardizing-pysyft-protocol.md). The approach taken
here is aimed at transmitting Protobuf over Websockets, and doesn't
attempt to define gRPC services--see the [RFC PR discussion](https://github.com/OpenMined/rfcs/pull/6) for the
reasoning behind that. See also proposed change 5 in the [serde
refactoring issue](https://github.com/OpenMined/PySyft/issues/2654).

This PR includes schemas for the existing messages in [message.py](https://github.com/OpenMined/PySyft/blob/master/syft/messaging/message.py) and
defines types where necessary to capture the content of those messages.
It includes a schema for Operation, which hopefully will be sufficiently
flexible to cover nearly all of the current uses of Operation in the PySyft
test suite. (I enumerated those uses by adding logging in Operation's
`__init__` and capturing the argument types. One known case that isn't
covered is nested tuples in the args, which I haven't sorted out yet.)

Schemas for Tensorflow types will be broadly similar and will make heavy
use of existing framework serialization (like the Torch types here do.) I
plan to add those in a future commit.

karlhigley commented on pysyft/types/syft/v1/arg.proto:
```diff
@@ -0,0 +1,27 @@
+        float arg_float = 3;
+        string arg_string = 4;
+
+        // TODO: Decide if we can live without VirtualWorker and just use int32 here
```
The VirtualWorker type is needed, because the `force_full_simplification` option can cause the VirtualWorker to be serialized with its tensors.

karlhigley commented on pysyft/types/torch/v1/tensor.proto:
```diff
@@ -0,0 +1,23 @@
+
+package pysyft.types.torch.v1;
+
+message Tensor {
```
Although the tests seem to pass around `torch.Tensor` a lot, we'll also need a serialized tensor schema for Tensorflow and maybe a generic Syft tensor schema too.

karlhigley commented:
Also need to write schemas for:

- [x] `syft.workers.Base`
- [x] `syft.messaging.plan.Plan`
- [x] `syft.messaging.Protocol`
- [x] `syft.generic.pointers.object_wrapper.ObjectWrapper`
- [x] `syft.exceptions.GetNotPermittedError`
- [x] `syft.exceptions.ResponseSignatureError`

vvmnnnkv commented on pysyft/types/torch/v1/tensor.proto:
```diff
@@ -0,0 +1,23 @@
+        int32 id_int = 1;
+        string id_str = 2;
+    }
+    bytes bin = 3;
```
Currently there may be 2 variants of `bin`, depending on `serializer`:
 * torch - pytorch serializer that makes binary blob. It's fast but useless for cross-language messages because it's very python-specific serialization, you need python to decode it.
 * all - makes tuple of shape, type, and array of values. This was meant as most generic way to send tensor

karlhigley replied:
Okay, cool, that confirms the suspicion I mentioned above that there’s a need for a more generic tensor representation.

vvmnnnkv replied:
Here's how generic serialization result looks like:
https://github.com/OpenMined/PySyft/pull/2762/commits/e6063f51dd74b825760e31585ec715d5c4f27f31#diff-89f20dd40f2dbc5ece5509b332de8394R531

I suspect TF will have its own tensor type. Not sure if it's better to join everything in one tensor type with `bin` field being `oneof` or it's better to have separate per-framework tensor types and then generic tensor type for cross-language/framework messaging.

karlhigley replied:
Thanks! I was leaning toward the latter. This `Tensor` schema just covers the Torch-specific serialization; I'll add a schema to match the generic serialization shortly.

karlhigley replied:
I take that back; l see that I've conflated these two options in this schema. This is closer to the generic Tensor with various serialization options, so let's give that a try and if it doesn't work out, we can always split it out into separate types.

karlhigley committed:
Add a dependency on the `protobuf` package

karlhigley committed:
Treat `Tensor` as a generic Syft type with various serializers

This schema looks like it may sufficiently general to cover Torch, TF,
and Syft tensors. We can split them back apart later if that turns out
not to be the case.

karlhigley committed:
Rename and flesh out the `Worker` type

`VirtualWorker` is usually used in the test suite, but this type actually
reflects serialization defined in `BaseWorker`. Let's just call it
`Worker` for now until there's a reason to specialize in particular types
of workers.

karlhigley committed:
Fix references to `Tensor` proto

karlhigley committed:
Create and apply an `Id` proto

This hadn't mattered in the previous schema definitions, the schema
for `State` will contain a list of ids. Since ids are `oneof`s and those
can't be repeated, ids will need a wrapper type.

karlhigley committed:
Add `Plan`, `Procedure`, `State` schemas

karlhigley committed:
Add Protocol schema

karlhigley committed:
Add `ObjectWrapper` schema

karlhigley committed:
Align the directory/file/package structure more closely with PySyft

karlhigley committed:
Add `Exceptions` proto

karlhigley commented:
At this point, I think I've covered the majority of stuff that gets serialized. One that I'm pretty sure I missed is `numpy.ndarray`—I'm not sure what's trying to serialize that yet, but I'll add it once I figure out where it's coming from.

karlhigley committed:
Add the Python stubs generated by `protoc` to the Python package

karlhigley commented:
There's a wrinkle I don't quite understand with the operations in a `Procedure` which seem to already be [simplified into Python primitives](https://github.com/OpenMined/PySyft/blob/master/syft/messaging/plan/procedure.py#L21). I assumed in the [Procedure proto](https://github.com/OpenMined/proto/pull/12/commits/88a9cf7d779c4d46e03fc5bb1c9d7d85069ea2a0#diff-3cd26b1efed9d06f448c945205fc3400R9) that it will somehow be possible to capture them as Protobuf objects instead, but I'm not entirely sure how that's going to work yet.

karlhigley commented:
Thus far, I've been working from the types exercised by the test suite (captured via logging in the serdes while the test suite runs.) Taking a different tack, searching the PySyft repo for all definitions of `simplify()` turns up some additional types, which suggest that the test suite doesn't exercise all of the serialization. Not sure how big an issue that is, but here's the list of all the places that define `simplify()` that aren't covered by the tests:
- [x] syft/generic/pointers/object_pointer.py
- [x] syft/generic/pointers/pointer_protocol.py
- [x] syft/generic/pointers/multi_pointer.py
- [x] syft/generic/pointers/pointer_plan.py
- [x] syft/frameworks/torch/tensors/interpreters/promise.py
- [x] syft/frameworks/torch/tensors/interpreters/additive_shared.py
- [x] syft/frameworks/torch/tensors/interpreters/autograd.py
- [x] syft/frameworks/torch/tensors/interpreters/private.py
- [x] syft/frameworks/torch/tensors/interpreters/precision.py
- [x] syft/frameworks/torch/tensors/interpreters/paillier.py (all attributes covered by generic tensor)
- [x] syft/frameworks/torch/tensors/interpreters/crt_precision.py (all attributes covered by precision tensor)
- [x] syft/frameworks/torch/tensors/decorators/logging.py (all attributes covered by generic tensor)
- [x] syft/messaging/promise.py (`simplify` defined but not yet implemented)
- [x] syft/federated/train_config.py

karlhigley committed:
Put wrappers and pointers together in a package (as prep for more)

karlhigley committed:
Convert `PointerTensor` into a generic `Pointer` class

The fields on the pointer types are all subsets of the same set of
fields. Capturing them as separate schemas might offer some additional
type safety, but feels unnecessarily repetitive and prone to maintenance
issues. They can be split later if there's a concrete reason to do so,
but seems like overkill for now. (Adding a field to capture the object
type of the pointer as a substitute.)

karlhigley committed:
Add `MultiPointer` schema

The typical (and AFAIK only) usage of this would be to bundle a list of
tensor pointers, but it's theoretically capable of being used for any
other type that Pointers are capable of referencing (Protocols, Plans,
etc.)

karlhigley committed:
Add `TrainConfig` schema

I'm assuming that the optimizer arguments are going to be numerical,
but that could be a mistaken assumption.

karlhigley committed:
Move `Tensor` schema into `generic` package

karlhigley committed:
Move `Worker` schema into `workers` package

karlhigley committed:
Fix `Tensor` references

karlhigley committed:
Add Torch tensor interpreter schemas

karlhigley committed:
Rebuild the Python stubs

karlhigley commented:
The remaining issues I know of are:
* _How to serialize numpy arrays._ Still not sure where they're coming from.
* _How to serialize Tensor sub-classes (i.e. the interpreters in the list above.)_ Protobuf doesn't have inheritance like mechanisms, so the options I know of are wrapper types or fields of type `Any`. I like the former but could be convinced to try the latter.
* _If and how to serialize Python functions (see the TODOs in the diff.)_ This appears to be a thing the current code does, but I wouldn't expect any of the JS/Android/iOS workers to be able to execute them. Possible I'm misreading, misunderstanding, or having a failure of imagination though.

karlhigley commented:
@cereallarceny @vvmnnnkv @iamtrask This could probably use some more eyes. I'm thinking that it might make sense to merge this once the known issues are resolved, just so there isn't a large diff hanging out on a branch for a long time. (Since nothing depends on it yet, should be relatively safe.) We'll definitely turn up further issues with these schemas while integrating them into the PySyft serialization, but I figure those can be addressed in further PRs. Thoughts?

iamtrask commented:
> The remaining issues I know of are:
> 
> * _How to serialize numpy arrays._ Still not sure where they're coming from.
> * _How to serialize Tensor sub-classes (i.e. the interpreters in the list above.)_ Protobuf doesn't have inheritance like mechanisms, so the options I know of are wrapper types or fields of type `Any`. I like the former but could be convinced to try the latter.
> * _If and how to serialize Python functions (see the TODOs in the diff.)_ This appears to be a thing the current code does, but I wouldn't expect any of the JS/Android/iOS workers to be able to execute them. Possible I'm misreading, misunderstanding, or having a failure of imagination though.

- Numpy arrays might have their own protobuf spec somewhere... but if you don't support this in V1 that's ok. I don't think it's critical functionality at the moment. (i.e., it's experimental)
- This part is super important. You can look in the simplifiers and detailers to see how things are currently simplified into simple Python objects (which protobuf should have no problem simplifying). However, a better option might be to only write protobuf objects for objects which are already simplified?
- We 100% want to prevent this from happening. The ability to execute arbitrary python functions on the other machine is a major security vulnerability.

karlhigley commented:
> This part is super important. You can look in the simplifiers and detailers to see how things are currently simplified into simple Python objects (which protobuf should have no problem simplifying). However, a better option might be to only write protobuf objects for objects which are already simplified?

I've been using the simplifiers to figure out what goes in each message, and sometimes that involves tracing backward through the code to figure out what types the fields should have. I'm not trying to encode the simplified output as Protobuf, but rather capturing the same fields in schemas that correspond relatively directly to the Python classes before simplification. (This maintains type information and avoids needing to keep a list of type constants.)

If we were trying to encode the simplified forms in Protobuf, we'd end up fighting the (useful) constraints of Protobuf to approximate a schema-less serialization approach, in which case we could just stay with msgpack.

karlhigley commented:
> * _How to serialize Tensor sub-classes (i.e. the interpreters in the list above.)_ Protobuf doesn't have inheritance like mechanisms, so the options I know of are wrapper types or fields of type Any. I like the former but could be convinced to try the latter.

The issue with sub-classes of `AbstractTensor` is that they mostly all define their own simplification/serialization, but have highly overlapping sets of fields. I'm not sure it makes sense to directly translate them each into a Protobuf schema. (`CRTPrecisionTensor` has no fields that aren't on `FixedPrecisionTensor` for example.) 

In order to maintain the separate types without having to list out (and maintain) all the fields each time, a normal approach would be to lean on inheritance, but Protobuf doesn't have any inheritance-like mechanisms. So it raises the question of how to encode separate types with overlapping fields with minimal duplication.

Options I can think of are:
* Accept the duplication and use a separate schema for each tensor sub-class. Maybe not ideal from a maintenance perspective, but easy enough to do.
* Add an enumeration in one or more of the tensor schemas to capture type information for multiple sub-classes. It's sort of a work around, but I suppose it could work.
* Add an `Any` field in one or more of the tensor schemas to capture the extra fields. This would make workers responsible for decoding the types of the additional fields, and knowing all the types of what can go in that field. Not looking forward to this.

I have maybe talked myself into doing the simplest thing and accepting the duplication, but definitely open to other thoughts.

karlhigley commented:
> Numpy arrays might have their own protobuf spec somewhere...

[This repo](https://github.com/xainag/numproto) shows one option for how to serialize them.

iamtrask commented:
So I have a personal coding preference for avoiding duplication EXCEPT when it's going to make the codebase significantly more complex because of un-supported abstractions. 

Aka, while bugs from duplicated code are a pain, at the moment the bigger pain is onboarding new programmers to the (Already very complex) PySyft codebase. 

My semi-confident vote is for the simple option.

karlhigley committed:
Add schemas for `PaillierTensor`, `CRTPrecisionTensor`, `LoggingTensor`

karlhigley commented:
> We 100% want to prevent this from happening. The ability to execute arbitrary python functions on the other machine is a major security vulnerability.

@iamtrask For reasons I'm not entirely clear on, [test_federated_client.py seems to be doing this in `test_fit`](https://github.com/OpenMined/PySyft/blob/master/test/federated/test_federated_client.py#L111) but not `test_evaluate`. I can't see any difference between the two that seems like it would lead to that behavior, so I'm going to ignore this for now and see if it crops up again later. (Maybe it has to do with the parameterization?)

karlhigley committed:
Add Python stubs for `AbstractTensor` sub-classes

karlhigley committed:
Fix import paths in Python package by rearranging directories

cereallarceny commented:
@karlhigley This looks like one hell of a PR - love it.  I'm talking a bit out of my expertise here, but let me just ask one question: is there any way to automate the creation of these proto files? The whole point of the Proto repo is to serve as a self-generating codebase that is refreshed with every new release of PySyft... not something that's edited by hand.  What can be done here if this process isn't already updated?

It may also be good to have @justin1121 and @jvmancuso check out this PR. They put together the RFC that includes protobuf in the first place.

karlhigley commented:
The proto schema files themselves pretty much have to be hand-written, since they define what the protocol is. The Python/JS/Java/Swift stubs are generated by the `protoc` compiler, so there is some automation in the pipeline.

Instead of PySyft being the source of truth and this repo being generated automatically, this repo becomes the source of truth and the parts that PySyft and the workers need get generated automatically from it.

karlhigley committed:
Add a `Message` wrapper schema w/ stubs

karlhigley commented:
Note: We'll need to account for OpenMined/PySyft#2684 (string types) when it gets merged.

karlhigley committed:
Fix typo in message proto

karlhigley committed:
Move the Protobuf schemas to account for the library name change

karlhigley committed:
Change the Protobuf package name to match the library name change

karlhigley committed:
Re-generate the stubs with the new library name

karlhigley commented:
This PR has been adjusted for the change from `proto` to `syft-proto`. Probably still some more changes coming as work through integrating it into PySyft.

karlhigley committed:
Update the Python stubs for the name change

karlhigley committed:
Update integer `Id`s to 64-bit

karlhigley committed:
Add Protobuf `Empty` type as possible contents of `SyftMessage`

Since communications are currently blocking, this handles the case where
a response has to be returned so the original sender awaiting a response
can move on. In Python, this looks like a `None` instead of an object
from a sub-class of Message.

karlhigley committed:
Update the type of `Operation.return_ids` to accommodate 64-bit ints

karlhigley committed:
Change `Operation` schema's `owner` field to a `Pointer`

karlhigley commented:
This PR is being split into smaller pieces so that schemas and translation can be tested end-to-end as they're added. The discussion here was helpful, and I'll continue to re-use and adapt the work from this PR as each chunk of serialization is developed.
